sdcm/nemesis_registry.py

```diff
@@ -0,0 +1,161 @@
+import ast
+import inspect
+import re
+from functools import lru_cache
+from typing import List, Optional, TypeVar, Callable
+
+from sdcm.utils.ast_utils import BooleanEvaluator
+
+DISRUPT_PATTERN = re.compile(r"self\.(?P<method_name>disrupt_[0-9A-Za-z_]+?)\(.*\)", flags=re.MULTILINE)
+SourceType = TypeVar("SourceType")
+DisruptMethod = Callable[[], None]
+
+
+@lru_cache
+def get_disrupt_method_from_class(nemesis_cls):
+    method_name = DISRUPT_PATTERN.search(inspect.getsource(nemesis_cls.disrupt))
+    if method_name:
+        return method_name.group("method_name")
+
+
+class NemesisRegistry:
+    def __init__(self, base_class: SourceType, kubernetes=False, excluded_list=None):
+        super().__init__()
+        self.base_class = base_class
+        self.kubernetes = kubernetes
+        self.excluded_list = excluded_list or []
+
+    def filter_subclasses(self, list_of_nemesis: List[SourceType], logical_phrase: str | None = None) -> List[SourceType]:
+        """
+        It apply 'and' logic to filter,
+            if any value in the filter does not match what nemeses have,
+            nemeses will be filtered out.
+        """
+        nemesis_subclasses = []
+
+        evaluator = BooleanEvaluator()
+        if logical_phrase:
+            expression_ast = ast.parse(logical_phrase, mode="eval")
+
+        for nemesis in list_of_nemesis:
+            if nemesis in self.excluded_list:
+                continue
+            evaluator.context = dict(**nemesis.__dict__, **{nemesis.__name__: True})
+            if logical_phrase and "disrupt_" in logical_phrase and (method_name := get_disrupt_method_from_class(nemesis)):
+                # if the `logical_phrase` has a method name of any disrupt method
+                # we look it up for the specific class and add it to the context
+                # so we can match on those as well
+                # example: 'disrupt_create_index or disrupt_drop_index'
+                evaluator.context[method_name] = True
+            if (logical_phrase and evaluator.visit(expression_ast)) or not logical_phrase:
+                nemesis_subclasses.append(nemesis)
+        return nemesis_subclasses
+
+    def get_subclasses(self, logical_phrase: str | None = None) -> List[SourceType]:
+        """Collects all known subclasses of Nemesis class"""
+        tmp = self.base_class.__subclasses__()
+        subclasses = []
+        while tmp:
+            for nemesis in tmp.copy():
+                subclasses.append(nemesis)
+                tmp.remove(nemesis)
+                tmp.extend(nemesis.__subclasses__())
+        return self.filter_subclasses(subclasses, logical_phrase)
+
+    def get_list_of_methods_by_flags(  # noqa: PLR0913
```

Comment:
I hate those functions quite a lot, cause they force us to define the flags multiple times

flag should be defined in one place (mostly by conversion, for now), and documented there.

this repetition is really pointless